staking/contracts/Staking.sol

```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
+            LibBrokenLine.LineData memory lineData = locks[account].locked.initiatedLines[id[i]];
+            (uint residue,) = locks[account].locked.remove(id[i], blockTime);
+
+            require(token.transfer(migrateTo, residue), "Failure while transferring in staking migration");
+            locks[account].amount = locks[account].amount.sub(residue);
+
+            locks[delegate].balance.remove(id[i], blockTime);
+            totalSupplyLine.remove(id[i], blockTime);
+            try nextVersionStake.initiateData(id[i], lineData, account, delegate) {
+            } catch {
+                revert("Contract not support or contain an error in interface INextVersionStake");
+            }
         }
-        //todo Lock delete
     }
 
-    function getStake(uint amount, uint slope, uint cliff) internal returns (uint){
-        return amount;
+    function verification(address account, uint id, uint newAmount, uint newSlope, uint newCliff, uint toTime) internal view {
+        require(account != address(0), "Line with id already deleted");
+        require(newAmount > 0, "Lock amount Rari mast be > 0");
+        require(newCliff <= TWO_YEAR_WEEKS, "Cliff period more, than two years");
+        uint period = newAmount.div(newSlope);
+        require(period <= TWO_YEAR_WEEKS, "Slope period more, than two years");
+        uint end = toTime.add(newCliff).add(period);
+        LibBrokenLine.LineData memory lineData = locks[account].locked.initiatedLines[id];
+        LibBrokenLine.Line memory line = lineData.line;
+        uint oldPeriod = line.bias.div(line.slope);
+        uint oldEnd = line.start.add(lineData.cliff).add(oldPeriod);
+        require(oldEnd <= end, "New line period stake too short");
     }
 
-    function increaseUnlockTime(uint lockId, uint period) public {
+    function removeLines(uint id, address account, address delegate, uint newAmount, uint toTime) internal {
+        uint bias = locks[account].locked.initial.bias;
+        uint balance = locks[account].amount.sub(bias);
+        (uint residue,) = locks[account].locked.remove(id, toTime);
+        //original: (uint residue, uint slope), but slope not need here
+        require(residue <= newAmount, "Impossible to restake: less amount, then now is");
+
+        uint addAmount = newAmount.sub(residue);
+        if (addAmount > balance) {//need more, than balance, so need transfer ERC20 to this
+            require(token.transferFrom(deposits[id].locker, address(this), addAmount.sub(balance)), "Failure while transferring");
+            locks[account].amount = locks[account].amount.sub(residue);
+            locks[account].amount = locks[account].amount.add(newAmount);
+        }
+        locks[delegate].balance.remove(id, toTime);
+        totalSupplyLine.remove(id, toTime);
+    }
 
+    function addLines(address account, address newDelegate, uint newAmount, uint newSlope, uint newCliff, uint blockTime) internal returns (uint) {
```

Comment:
this function looks like stake. Is it better to invoke addLines from stake?

Reply:
agree, do it